Update driver version to v5.7.9_35795.20191128

baselsayeh opened the pull request:
Works greatly on orange pi zero plus,
tested on armbian
lightly tested on openwrt

Changelog:
v5.7.9_35795.20191128:
- Hotfix P2P NoA buffer potentially overflow
- Fix Adaptivity blocks scan issue
- Fix can't trigger disconnect while change of AP channel when scan triggers frequently
- Fix SSID of beacon can't follow hidden SSID mode before second issue_beacon() is called
- Fix non-target(w/o scan_ssid) scanned network can't be
- updated with current IE after changing to hidden
- Support ap_isolate=1(p2p_intra_bss=0) option
- Add DFS triggered by scan with interference info
- IOT: Degrade bandwidth to HT20 if AP declare HT40 without 2nd channel setting

v5.7.8_34044.20190614:
- Improve TX path efficiency
- Improve long distance performance

v5.3.16_32695.20190327:
- Fix wrong internal band status

v5.3.15_32624.20190320:
- Fix no DIG when enter LPS_LCLK

v5.3.12_28613.20180703:
- Support 802.11s MESH mode

Full change log is available at: [Web archive link](https://archive.org/download/rtl-8189-fs-linux-v-5.7.9-35795.20191128/ReleaseNotes.pdf)

Full driver package is at: [Web archive link](https://archive.org/details/rtl-8189-fs-linux-v-5.7.9-35795.20191128)

clnhub committed:
Add kernel 5.12(+) (GRO_DROP removed) support

CGarces commented:
Thanks for the PR.
I need to check the error for 4.14,about the other compilation errors...

Can you please add this patch to your code? 
https://github.com/Mange/rtl8192eu-linux-driver/pull/233

Also I think that https://github.com/jwrdegoede/rtl8189ES_linux/pull/71/commits/61887e8c93f24a5625a42ddc84b2fed38fb06d74 should be merged with the commit that introduce the error instead of create a new commit.

baselsayeh commented:
> Thanks for the PR. I need to check the error for 4.14,about the other compilation errors...
> 
> Can you please add this patch to your code? [Mange/rtl8192eu-linux-driver#233](https://github.com/Mange/rtl8192eu-linux-driver/pull/233)
> 
> Also I think that [61887e8](https://github.com/jwrdegoede/rtl8189ES_linux/commit/61887e8c93f24a5625a42ddc84b2fed38fb06d74) should be merged with the commit that introduce the error instead of create a new commit.

Done
Removed the extra "small fixes" patch, and added the GRO_DROP patch

Edit: Oops, fixed a minor compilation issue in the old "use #ifdef CONFIG_OF in rtw_ieee80211" commit. Now it should compile correctly

CGarces commented:
Looks good for me

Ping @martinayotte or @igorpecovnik maybe they can test on ArmBian builds before merge.

I'll merge the next week if I not receive negative comments.

CGarces commented:
@baselsayeh where this code comes from? there is a changelog?

baselsayeh commented:
> @baselsayeh where this code comes from? there is a changelog?

Sorry for not mentioning this in the first place
I mainly found it on
[this website](https://gitee.com/logic_2050/RTL8189FTV-aml/tree/master/rtl8189FS)
and
[this repo](https://github.com/hardkernel/buildroot_platform_hardware_wifi_realtek_drivers_8189ftv/tree/aml64_buildroot_master_c4)

Dont know the exact changelog, it mainly says "update driver to support android p"
Also found that some wifi mesh code was added, mainly found in the core/mesh dir

baselsayeh commented:
Hello @CGarces 
I found a newer driver (along with a changelog, i found the driver on [csdn](https://download.csdn.net/download/jurrah/12683297?spm=1003.2122.3001.6634.6) )
Do i close this PR and create a new one? Or revert and force-push patches?

CGarces commented:
Feel free to force push

baselsayeh committed:
Update version to v5.7.9_35795.20191128

baselsayeh committed:
fixes before 5.1

baselsayeh committed:
support for getting mac-address from devicetree

This patch got removed when the source files
got updated to the newer version

baselsayeh committed:
5.1-5.17 fixes

baselsayeh committed:
use #ifdef CONFIG_OF in rtw_ieee80211

This patch fixes the compilation on x86
kernels who does not enable CONFIG_OF

baselsayeh committed:
Disable CONFIG_RTW_NAPI

It does not play well with newer kernels
for some reason, frames do not arrive to the kernel with this enabled, so disable it

baselsayeh commented:
For anyone's reading: i also uploaded an updated rtl8189ES driver to the web archive, [link](https://archive.org/details/rtl-8189-es-linux-v-5.8.9-35085.20191003)
If anyone is ready to test the driver for rtl8189ES, contact me and i will make a new PR

baselsayeh commented:
Hello @CGarces 
when will it get merged? i already submitted a PR to armbian that will fix any compilation errors resulted from the update.

CGarces commented:
Opps, sorry.

emanuele-f commented:
> For anyone's reading: i also uploaded an updated rtl8189ES driver to the web archive, [link](https://archive.org/details/rtl-8189-es-linux-v-5.8.9-35085.20191003) If anyone is ready to test the driver for rtl8189ES, contact me and i will make a new PR

@baselsayeh I'd like to test the updated rtl8189ES driver on my [nanopi r1s h3](http://wiki.friendlyelec.com/wiki/index.php/NanoPi_R1S-H3) running armbian. The current driver has a bug which causes 100% cpu usage and very low throughput (~9mbit/s) while running as a wifi hotspot, so I'd like to check if the updated driver fixes this.

![img](https://user-images.githubusercontent.com/5488003/166944466-c4457cd4-f09a-4c90-b10e-3fe42137be9a.png)

Here are more details for the current driver:

```
$ lshw -C network
configuration: broadcast=yes driver=rtl8189es driverversion=5.15.35-sunxi multicast=yes wireless=IEEE 802.11bg
  *-network:2

$ modinfo 8189es | grep version
version:        v4.3.18.1_15373.20151005-20181020
srcversion:     2E7692D0CCD9BEFA8A26D08
parm:           rtw_chip_version:int
```

Please let me know how I can apply your patch

igorpecovnik commented:
> Please let me know how I can apply your patch

Armbian nightly builds should have this code. (armbian-config -> switch to nightly) + reboot.

emanuele-f commented:
> Armbian nightly builds should have this code. (armbian-config -> switch to nightly) + reboot.

The output above was collected on the latest master built from github. In my understanding this PR currently only applies to `rtl8189fs` but not to `rtl8189es`.

Ps. indeed the `rls8189fs` module is new one:
```
modinfo 8189fs | grep version
version:        v5.7.9_35795.20191128
```

baselsayeh commented:
> > Armbian nightly builds should have this code. (armbian-config -> switch to nightly) + reboot.
> 
> The output above was collected on the latest master built from github. In my understanding this PR currently only applies to `rtl8189fs` but not to `rtl8189es`.
> 
> Ps. indeed the `rls8189fs` module is new one:
> 
> ```
> modinfo 8189fs | grep version
> version:        v5.7.9_35795.20191128
> ```

That is correct.

I made the necessary patches in my fork 

you should be able to compile the module via
```
git clone https://github.com/baselsayeh/rtl8189ES_linux
cd rtl8189ES_linux
make ARCH=arm CROSS_COMPILE= KSRC=/lib/modules/`uname -r`/build
rmmod 8189es
insmod ./8189es.ko
```
but make sure first to download the kernel headers in armbian-config

and try it. if everything works, i will create a new PR (most importantly, test AP and client, see if both works)

emanuele-f commented:
@baselsayeh module built, it's running well, stats below. `armbian-config` did not install the correct headers (probably because I built armbian from source) but I solved this by manually installing the [headers from the repo](https://imola.armbian.com/beta/pool/main/l/linux-5.15.35-sunxi/linux-headers-current-sunxi_22.05.0-trunk.0067_armhf.deb).

Here is the built module [8189es.ko.tar.gz](https://github.com/jwrdegoede/rtl8189ES_linux/files/8639978/8189es.ko.tar.gz)

```
$ uname -r
5.15.35-sunxi

$ modinfo 8189es.ko | grep version
version:        v5.8.9_35085.20190919
```

A small correction to my previous test, the old kernel module performs better (on pair of this new one) if 802.11n is enabled, I run it in g mode since I thought it was not supported, but figured out that `sytemctl restart hostap` sometimes fails, so a second restart is needed to properly start the AP. Nevertheless, the 100% cpu usage is still a problem with the old module, which is now fixed with this new module.

## AP mode

Hostap settings:

```
interface=wlan0
hw_mode=g
channel=3
driver=nl80211
ieee80211n=1
ht_capab=[HT20][SHORT-GI-20]
wmm_enabled=1
```

The 100% cpu usage problem of the RTWHALXT process during the speedtest is now solved. The cpu usage now is below 15%

![ht2](https://user-images.githubusercontent.com/5488003/167130952-c9a411c6-3a41-424c-a0c0-54e0ca968f16.png)

```
$ iwconfig

wlan0     IEEE 802.11bgn  ESSID:"xxxxxxx"  Nickname:"<WIFI@REALTEK>"
          Mode:Master  Frequency:2.422 GHz  Access Point: C4:3A:35:xx:xx:xx
          Bit Rate:72.2 Mb/s   Sensitivity:0/0  
          Retry:off   RTS thr:off   Fragment thr:off
          Encryption key:off
          Power Management:off
          Link Quality=1/100  Signal level=1/100  Noise level=0/100
          Rx invalid nwid:0  Rx invalid crypt:0  Rx invalid frag:0
          Tx excessive retries:0  Invalid misc:0   Missed beacon:0
```

Wifi clients connect and use the Internet without problems.
`speedtest` reports a download a rate which ranges in 25-40 mbps. Upload rate is not meaningful since its capped by my adsl. Iperf in udp mode reports a max bandwidth of 50 mbps both in upload and download.

## Client mode

Client mode also works correctly.

The [official wifi rates](https://wiki.friendlyelec.com/wiki/index.php/NanoPi_R1S-H3#Hardware_Spec) are 76.6 mbps for download and 84 mbps for upload. With `speedtest-cli` I achieved 62 mbps in download. With iperf, I get 100 mbps for download (which is higher than the official rate) and 75 mbps for upload (lower than official rate).

baselsayeh commented:
> @baselsayeh module built, it's running well, stats below. `armbian-config` did not install the correct headers (probably because I built armbian from source) but I solved this by manually installing the [headers from the repo](https://imola.armbian.com/beta/pool/main/l/linux-5.15.35-sunxi/linux-headers-current-sunxi_22.05.0-trunk.0067_armhf.deb).
> 
> Here is the built module [8189es.ko.tar.gz](https://github.com/jwrdegoede/rtl8189ES_linux/files/8639978/8189es.ko.tar.gz)
> 
> ```
> $ uname -r
> 5.15.35-sunxi
> 
> $ modinfo 8189es.ko | grep version
> version:        v5.8.9_35085.20190919
> ```
> 
> A small correction to my previous test, the old kernel module performs better (on pair of this new one) if 802.11n is enabled, I run it in g mode since I thought it was not supported, but figured out that `sytemctl restart hostap` sometimes fails, so a second restart is needed to properly start the AP. Nevertheless, the 100% cpu usage is still a problem with the old module, which is now fixed with this new module.
> 
> ## AP mode
> Hostap settings:
> 
> ```
> interface=wlan0
> hw_mode=g
> channel=3
> driver=nl80211
> ieee80211n=1
> ht_capab=[HT20][SHORT-GI-20]
> wmm_enabled=1
> ```
> 
> The 100% cpu usage problem of the RTWHALXT process during the speedtest is now solved. The cpu usage now is below 15%
> 
> ![ht2](https://user-images.githubusercontent.com/5488003/167130952-c9a411c6-3a41-424c-a0c0-54e0ca968f16.png)
> 
> ```
> $ iwconfig
> 
> wlan0     IEEE 802.11bgn  ESSID:"xxxxxxx"  Nickname:"<WIFI@REALTEK>"
>           Mode:Master  Frequency:2.422 GHz  Access Point: C4:3A:35:xx:xx:xx
>           Bit Rate:72.2 Mb/s   Sensitivity:0/0  
>           Retry:off   RTS thr:off   Fragment thr:off
>           Encryption key:off
>           Power Management:off
>           Link Quality=1/100  Signal level=1/100  Noise level=0/100
>           Rx invalid nwid:0  Rx invalid crypt:0  Rx invalid frag:0
>           Tx excessive retries:0  Invalid misc:0   Missed beacon:0
> ```
> 
> Wifi clients connect and use the Internet without problems. `speedtest` reports a download a rate which ranges in 25-40 mbps. Upload rate is not meaningful since its capped by my adsl. Iperf in udp mode reports a max bandwidth of 50 mbps both in upload and download.
> 
> ## Client mode
> Client mode also works correctly.
> 
> The [official wifi rates](https://wiki.friendlyelec.com/wiki/index.php/NanoPi_R1S-H3#Hardware_Spec) are 76.6 mbps for download and 84 mbps for upload. With `speedtest-cli` I achieved 62 mbps in download. With iperf, I get 100 mbps for download (which is higher than the official rate) and 75 mbps for upload (lower than official rate).

Great. I will make a new PR, but need to check if the armbian build system can build the new driver without any problems.

And for the Hotspot, I think you can get higher speeds if you enable hostapd to use 40MHz (it will refuse if it finds any other AP working in the same channel as hostapd).
